### PR TITLE
Implement query function

### PR DESCRIPTION
**Pull request by kyracq**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,96 @@
...
+      TimeRange validRange = TimeRange.fromStartEnd(validRangeStart, validRangeEnd, false);
+      // Add to results if it's long enough
+      if (validRange.duration() >= duration) {
+        results.add(validRange);
+      }
+    }
+
+    invalidRangeStart = invalidRanges.get(0).start();
+    minInvalidRangeEnd = invalidRanges.get(0).end();
+    for (int i = 1; i < invalidRanges.size(); i++) {
```
Comment (shulijones) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: Could add another comment here describing your general strategy in this for loop

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,96 @@
...
+    }
+
+    ArrayList<TimeRange> invalidRanges = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      Set<String> intersection = new HashSet<String>(attendees);
+
+      // If there are attendees in common between the request and this event
+      if(!Collections.disjoint(attendees, event.getAttendees())) {
+        // Requested meeting can't be at same time as this event
+        invalidRanges.add(event.getWhen());
```
Comment (shulijones) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: I like the way you just work with TimeRange so you don't have to deal with Events.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,96 @@
...
+
+    invalidRangeStart = invalidRanges.get(0).start();
+    minInvalidRangeEnd = invalidRanges.get(0).end();
+    for (int i = 1; i < invalidRanges.size(); i++) {
+      if (invalidRanges.get(i).start() > minInvalidRangeEnd) {
+        // We've found an open time slot
+        validRangeStart = minInvalidRangeEnd;
+        validRangeEnd = invalidRanges.get(i).start();
+        TimeRange validRange = TimeRange.fromStartEnd(validRangeStart, validRangeEnd, false);
+        if(validRange.duration() >= duration) {
```
Comment (shulijones) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: space after if

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,96 @@
...
+      } else {
+        // We definitely won't find a valid range until at least the end of curr invalid range
+        minInvalidRangeEnd = Math.max(invalidRanges.get(i).end(), minInvalidRangeEnd);
+      }
+    }
+
+    // Add last block of day to results (if long enough)
+    if (minInvalidRangeEnd < TimeRange.END_OF_DAY) {
+      TimeRange validRange = TimeRange.fromStartEnd(minInvalidRangeEnd, TimeRange.END_OF_DAY, true);
+      if(validRange.duration() >= duration) {
```
Comment (shulijones) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: space after if

**Review by shulijones — APPROVED** [review]
LGTM! Seems really nice and clean.

**Review by otobin — APPROVED** [review]
LGTM!

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 5]
```diff
@@ -30,4 +30,24 @@ public void testGreeting() {
...
+
+    // Whitespace should be trimmed
+    Assert.assertEquals("Hello Ada", greeting);
+  }
+
+  @Test
+  public void testGreetingLettersOnly() {
+    Greeter greeter = new Greeter();
+
+    String greeting = greeter.greet("@#$&*#Ada");
```
Comment (anicr7) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: How about adding numbers here too?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,93 @@
...
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+  /**
+   * Given the currently booked events (time ranges and attendees) and
+   * a meeting request (duration and attendees), return a collection
+   * the of time ranges within which the meeting can be booked.
```
Comment (anicr7) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,93 @@
...
+    }
+
+    if (invalidRanges.isEmpty()) {
+      results.add(TimeRange.WHOLE_DAY);
+      return results;
+    }
+
+    Collections.sort(invalidRanges, TimeRange.ORDER_BY_START);
+    
+    int invalidRangeStart;
```
Comment (anicr7) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: Simpler to use one-line for this instead of 4 different lines.

**Review by anicr7 — CHANGES_REQUESTED** [review]

**Review by anicr7 — APPROVED** [review]

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 8]
```diff
@@ -30,4 +30,24 @@ public void testGreeting() {
 
     Assert.assertEquals("Hello Ada", greeting);
   }
+
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:33]: You should also think about what are called "edge cases" so for example what if you pass the empty string or a string with empty spaces only.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (nsri-google) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: I think the style guide recommends separate lines.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,93 @@
...
+    int minInvalidRangeEnd;
+    int validRangeStart;
+    int validRangeEnd;
+
+    // If first event doesn't start at start of day, add time range beginning
+    // at start of day
+    if (invalidRanges.get(0).start() != TimeRange.START_OF_DAY) {
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: invalidRanges.get(0) is being used multiple times so it might be useful to store it in a variable

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,93 @@
...
+      if (validRange.duration() >= duration) {
+        results.add(validRange);
+      }
+    }
+
+    invalidRangeStart = invalidRanges.get(0).start();
+    minInvalidRangeEnd = invalidRanges.get(0).end();
+    // Look for open time slots between invalid ranges
+    for (int i = 1; i < invalidRanges.size(); i++) {
+      if (invalidRanges.get(i).start() > minInvalidRangeEnd) {
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: invalidRanges.get(i) is being used multiple times so it might be useful to store it in a variable

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,93 @@
...
+      // Add to results if it's long enough
+      if (validRange.duration() >= duration) {
+        results.add(validRange);
+      }
+    }
+
+    invalidRangeStart = invalidRanges.get(0).start();
+    minInvalidRangeEnd = invalidRanges.get(0).end();
+    // Look for open time slots between invalid ranges
+    for (int i = 1; i < invalidRanges.size(); i++) {
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: since you only care about the value and not the index (i) you could use the for each loop version (see method 2 in https://www.geeksforgeeks.org/iterating-arraylists-java/).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,93 @@
...
+      if (invalidRanges.get(i).start() > minInvalidRangeEnd) {
+        // We've found an open time slot
+        validRangeStart = minInvalidRangeEnd;
+        validRangeEnd = invalidRanges.get(i).start();
+        TimeRange validRange = TimeRange.fromStartEnd(validRangeStart, validRangeEnd, false);
```
Comment (nsri-google) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The same sequence of steps is done in multiple places, maybe you should have a helper function for it.

**Review by nsri-google — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (kyracq) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: I did consider using a for each loop however I needed to start at index 1, so I thought I'd be simpler to use a regular for loop where I can set i to start at 1.

**Review by nsri-google — APPROVED** [review]
